File: tools/zebra_bridge.py
import json
import socket
from http.server import BaseHTTPRequestHandler, HTTPServer
from typing import List, Optional

try:
    import win32print
except ImportError:
    win32print = None


HOST = "0.0.0.0"
PORT = 9123
DEFAULT_PRINTER = "ZD220-203dpi ZPL"
ZEBRA_HINTS = ("zebra", "zd", "zdesigner", "zpl")


def get_local_ip() -> str:
    try:
        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as sock:
            sock.connect(("8.8.8.8", 80))
            return sock.getsockname()[0]
    except OSError:
        return "IP_DO_NOTEBOOK"


def list_printers() -> List[str]:
    if win32print is None:
        raise RuntimeError("Instale o pywin32: pip install pywin32")

    flags = win32print.PRINTER_ENUM_LOCAL | win32print.PRINTER_ENUM_CONNECTIONS
    return [printer[2] for printer in win32print.EnumPrinters(flags) if printer[2]]


def detect_zebra_printer() -> str:
    printers = list_printers()
    for printer in printers:
        lower_name = printer.lower()
        if any(hint in lower_name for hint in ZEBRA_HINTS):
            return printer

    default_printer = win32print.GetDefaultPrinter()
    if default_printer:
        return default_printer

    raise RuntimeError("Nenhuma impressora instalada foi encontrada")


def send_raw_zpl(printer_name: Optional[str], zpl: str) -> str:
    if win32print is None:
        raise RuntimeError("Instale o pywin32: pip install pywin32")

    printer_name = printer_name or detect_zebra_printer()
    handle = win32print.OpenPrinter(printer_name)
    try:
        job = win32print.StartDocPrinter(handle, 1, ("MyKids etiqueta", None, "RAW"))
        try:
            win32print.StartPagePrinter(handle)
            win32print.WritePrinter(handle, zpl.encode("utf-8"))
            win32print.EndPagePrinter(handle)
        finally:
            win32print.EndDocPrinter(handle)
    finally:
        win32print.ClosePrinter(handle)
    return printer_name


class ZebraBridgeHandler(BaseHTTPRequestHandler):
    def end_headers(self):
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.send_header("Access-Control-Allow-Methods", "POST, OPTIONS")
        super().end_headers()

    def do_OPTIONS(self):
        self.send_response(204)
        self.end_headers()

    def do_GET(self):
        if self.path not in ("/", "/health"):
            self.send_response(404)
            self.end_headers()
            return

        try:
            printer_name = detect_zebra_printer()
            printers = list_printers()
            ok = True
            error = None
        except Exception as exc:
            printer_name = None
            printers = []
            ok = False
            error = str(exc)

        self.send_response(200 if ok else 500)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(
            json.dumps(
                {
                    "ok": ok,
                    "service": "mykids-zebra-bridge",
                    "printerName": printer_name,
                    "printers": printers,
                    "printUrl": f"http://{get_local_ip()}:{PORT}/print",
                    "error": error,
                }
            ).encode("utf-8")
        )

    def do_POST(self):
        if self.path != "/print":
            self.send_response(404)
            self.end_headers()
            return

        try:
            length = int(self.headers.get("Content-Length", "0"))
            payload = json.loads(self.rfile.read(length).decode("utf-8"))
            zpl = str(payload.get("zpl", ""))
            printer_name = str(payload.get("printerName") or "").strip() or None

            if not zpl.strip():
                raise ValueError("Campo zpl vazio")

            printer_name = send_raw_zpl(printer_name, zpl)
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({"ok": True, "printerName": printer_name}).encode("utf-8"))
        except Exception as exc:
            self.send_response(500)
            self.send_header("Content-Type", "application/json")
            self.end_headers()
            self.wfile.write(json.dumps({"ok": False, "error": str(exc)}).encode("utf-8"))


if __name__ == "__main__":
    server = HTTPServer((HOST, PORT), ZebraBridgeHandler)
    local_ip = get_local_ip()
    print(f"MyKids Zebra bridge rodando no notebook em http://127.0.0.1:{PORT}/print")
    print(f"Para usar pelo celular na mesma rede, configure no site: http://{local_ip}:{PORT}/print")
    print(f"Teste no navegador do celular: http://{local_ip}:{PORT}/health")
    try:
        print(f"Impressora detectada: {detect_zebra_printer()}")
    except Exception as exc:
        print(f"Impressora detectada: erro - {exc}")
    server.serve_forever()

File: tools/zebra_bridge.js
const crypto = require("crypto");
const fs = require("fs");
const http = require("http");
const os = require("os");
const path = require("path");
const { execFile } = require("child_process");

const HOST = "0.0.0.0";
const PORT = 9123;
const ZEBRA_HINTS = ["zebra", "zd", "zdesigner", "zpl"];

function runPowerShell(args) {
  return new Promise((resolve, reject) => {
    execFile("powershell.exe", ["-NoProfile", "-ExecutionPolicy", "Bypass", ...args], {
      windowsHide: true,
      timeout: 15000,
    }, (error, stdout, stderr) => {
      if (error) {
        reject(new Error((stderr || error.message || "Falha no PowerShell").trim()));
        return;
      }
      resolve(stdout);
    });
  });
}

async function listPrinters() {
  const stdout = await runPowerShell([
    "-Command",
    "Add-Type -AssemblyName System.Drawing; $default = (New-Object System.Drawing.Printing.PrinterSettings).PrinterName; [System.Drawing.Printing.PrinterSettings]::InstalledPrinters | ForEach-Object { [pscustomobject]@{ Name = [string]$_; Default = ([string]$_ -eq $default) } } | ConvertTo-Json -Compress",
  ]);
  if (!stdout.trim()) return [];
  const parsed = JSON.parse(stdout);
  return (Array.isArray(parsed) ? parsed : [parsed]).filter((printer) => printer && printer.Name);
}

async function detectPrinter() {
  const printers = await listPrinters();
  const hinted = printers.find((printer) => {
    const name = String(printer.Name).toLowerCase();
    return ZEBRA_HINTS.some((hint) => name.includes(hint));
  });
  if (hinted) return hinted.Name;

  const defaultPrinter = printers.find((printer) => printer.Default);
  if (defaultPrinter) return defaultPrinter.Name;

  throw new Error("Nenhuma impressora instalada foi encontrada");
}

function getLocalIp() {
  const interfaces = os.networkInterfaces();
  for (const entries of Object.values(interfaces)) {
    for (const entry of entries || []) {
      if (entry.family === "IPv4" && !entry.internal) return entry.address;
    }
  }
  return "127.0.0.1";
}

function sendJson(response, statusCode, body) {
  response.writeHead(statusCode, {
    "Content-Type": "application/json; charset=utf-8",
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "Content-Type",
    "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
    "Access-Control-Allow-Private-Network": "true",
  });
  response.end(JSON.stringify(body));
}

function readJson(request) {
  return new Promise((resolve, reject) => {
    let body = "";
    request.setEncoding("utf8");
    request.on("data", (chunk) => {
      body += chunk;
      if (body.length > 1024 * 1024) {
        request.destroy();
        reject(new Error("Payload muito grande"));
      }
    });
    request.on("end", () => {
      try {
        resolve(JSON.parse(body || "{}"));
      } catch {
        reject(new Error("JSON invalido"));
      }
    });
    request.on("error", reject);
  });
}

async function printZpl(printerName, zpl) {
  const resolvedPrinter = printerName || await detectPrinter();
  const zplPath = path.join(os.tmpdir(), `mykids-zpl-${crypto.randomUUID()}.zpl`);
  fs.writeFileSync(zplPath, zpl, "utf8");
  try {
    await runPowerShell([
      "-File",
      path.join(__dirname, "send_zpl.ps1"),
      "-PrinterName",
      resolvedPrinter,
      "-ZplPath",
      zplPath,
    ]);
  } finally {
    fs.rmSync(zplPath, { force: true });
  }
  return resolvedPrinter;
}

const server = http.createServer(async (request, response) => {
  const url = new URL(request.url || "/", `http://${HOST}:${PORT}`);

  if (request.method === "OPTIONS") {
    sendJson(response, 204, { ok: true });
    return;
  }

  if (request.method === "GET" && (url.pathname === "/" || url.pathname === "/health")) {
    try {
      const printers = await listPrinters();
      const printerName = await detectPrinter();
      sendJson(response, 200, {
        ok: true,
        service: "mykids-zebra-bridge",
        printerName,
        printers: printers.map((printer) => printer.Name),
        printUrl: `http://${getLocalIp()}:${PORT}/print`,
        error: null,
      });
    } catch (error) {
      sendJson(response, 200, {
        ok: false,
        service: "mykids-zebra-bridge",
        printerName: null,
        printers: [],
        printUrl: `http://${getLocalIp()}:${PORT}/print`,
        error: error.message,
      });
    }
    return;
  }

  if (request.method === "POST" && url.pathname === "/print") {
    try {
      const payload = await readJson(request);
      const zpl = String(payload.zpl || "");
      if (!zpl.trim()) throw new Error("Campo zpl vazio");
      const printerName = await printZpl(payload.printerName ? String(payload.printerName) : "", zpl);
      sendJson(response, 200, { ok: true, printerName });
    } catch (error) {
      sendJson(response, 500, { ok: false, error: error.message });
    }
    return;
  }

  sendJson(response, 404, { ok: false, error: "Nao encontrado" });
});

server.listen(PORT, HOST, () => {
  console.log(`MyKids Zebra bridge rodando em http://127.0.0.1:${PORT}/print`);
  console.log(`Teste: http://127.0.0.1:${PORT}/health`);
  console.log(`Rede local: http://${getLocalIp()}:${PORT}/print`);
});
